[MRG + 2] make check_array convert object to float.

amueller opened the pull request:
- [x] add tests
- [ ] ~~think about people passing strings as objects to SVM with callable kernel~~ Not too concerned about this any more. Maybe @larsmans can comment?
- [x] think about what to do with y

The problem with y is that for regressors we might want to convert y to float, whereas for classifiers we map any unique objects to ints anyhow, so object is fine. Fixes #4006, #3142 and #4055.

amueller commented:
Heisenbug in LogisticRegresion....

amueller commented:
It looks a bit like converting to float is not that big of an issue, as sting kernels already didn't work, see http://stackoverflow.com/questions/26391367/how-to-use-string-kernels-in-scikit-learn
It would be better to fix the issue, but that seems independent of the PR. It is a bit hard to test for features that we don't about.

amueller commented:
NaN/Inf error in ``OrthogonalMatchingPursuitCV`` on Python2.6.... I don't really understand....

amueller commented:
@ogrisel if you have time, I'd love to get your input on this, as this is a somewhat major behavior change.

amueller commented:
@GaelVaroquaux might also have an opinion about this (and will hopefully forgive me from distracting him from grand proposals)

agramfort commented:
no objection on my side

amueller commented:
travis seems unstable again :-/

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -267,11 +270,19 @@ def check_array(array, accept_sparse=None, dtype=None, order=None, copy=False,
+        if dtype == "numeric":
+            if getattr(array, "dtype", None) is np.dtype(object):
+                # if input is object, convert to float.
+                dtype = np.float
```
I would like to ban the use of `np.float` which is just an alias for Python `float`. Instead we should use the more explicit and platform-independent `np.float64`.

amueller replied:
Sounds like a good idea.

amueller replied:
Fixed. FYI:
> git grep "np.float" | grep -v float64 | grep -v float32 | grep py | wc -l
138

ogrisel replied:
Well, those other `np.float` will have to be dealt with separately ;)

ogrisel commented:
> think about people passing strings as objects to SVM with callable kernel Not too concerned about this any more. Maybe @larsmans can comment?

suggestion: when the kernel is an arbitrary callable, we don't enforce the input dtype to be numeric and keep it a pass-through (`dtype=None`).

amueller commented:
We could do it like that, but currently we don't have a working example of that afaik.

ogrisel commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1143,10 +1142,12 @@ def feature_importances_(self):
         self.n_classes_ = 1
-
+        if y.dtype is np.dtype(object):
+            y = y.astype(np.float)
```
One more `np.float` here.

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -305,8 +318,9 @@ def check_X_y(X, y, accept_sparse=None, dtype=None, order=None, copy=False,
 
-    dtype : string, type or None (default=none)
+    dtype : string, type or None (default="numeric")
         Data type of result. If None, the dtype of the input is preserved.
```
I would be more specific and say "Data type of the checked input data X" instead of "Data type of result".

ogrisel commented on sklearn/utils/validation.py:
```diff
@@ -328,6 +342,10 @@ def check_X_y(X, y, accept_sparse=None, dtype=None, order=None, copy=False,
 
+    y_numeric : boolean (default=False)
+        Whether to ensure that y has a numeric type. If dtype of y is object,
+        it is converted to float. Should only be used for regression algorithms.
```
float64

ogrisel commented:
This looks good to me. Any more comments @jnothman @arjoly @larsmans?

ogrisel commented:
I wonder if we should use `y_dtype=None` instead of `y_numeric=False` though. That would make it possible to implement finer control of the `y_dtype` precision level for some cython code. But maybe this is a YAGNI.

ogrisel commented:
This is somewhat related to the idiom discussed in https://github.com/scikit-learn/scikit-learn/pull/4136#discussion_r24176562 to add support for `dtype=[np.float64, np.float32]` for specific models (typically with cython code that would crash with integer inputs). For some models might want also want to have `y_dtype=[np.float64, np.float32]`.

amueller commented:
If we wanted ``y_dtype`` instead of  aboolean ``y_numeric`` the default value would not be ``None`` but the list of all np dtypes != ``object`` with ``np.float64`` as the first place, right?

amueller commented:
Oh, never mind, I misrembered what ``y_numeric`` does... if would be ``_dtype=[np.float64, np.float32]`` for the regression models, and ``None`` otherwise.

ogrisel commented:
> Oh, never mind, I misrembered what y_numeric does... if would be _dtype=[np.float64, np.float32] for the regression models, and None otherwise.

Yes, unless the regression model using a Cython or C/C++ implementation that expects double precision in which case we would do `y_dtype=np.float64` directly.

ogrisel commented:
There are still a bunch of `np.float` that should be `np.float64` in the lines touched by this PR.

amueller commented:
Damn, must have been rebase issues. Should be fixed now.

amueller commented:
As this is a somewhat major change, but will fix several issues, I feel it would be good to get a couple more reviews....

amueller commented:
Not sure if the OMP test failure is related.

amueller committed:
make check_array convert object to float.

fix dtype check, add test. unfriend all multi-output estimators on facebook.

try to fix what is happening to y (by doing nothing to y)

make test work...

Make everything accept object y or say "invalid label"

fix multioutput linear models

add test for sensible error message.

agramfort commented:
LGTM but you'll need to rebase.

jnothman commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -1146,7 +1146,8 @@ def feature_importances_(self):
     def _validate_y(self, y):
         self.n_classes_ = 1
-
+        if y.dtype is np.dtype(object):
```
If this logic is present in a classifier, it is problematic; strings should be valid as dtype object (e.g. this is default for a pandas.Series of string). If it's only relevant to regressors, it should probably not be in `Base`.

GaelVaroquaux replied:
I think that I would prefer if the check was written as:

y.dtype.kind == 'O'

amueller replied:
@jnothman this is how the code is factored. The regressor implementation is in Base, and overwritten in the classifier.

amueller replied:
I could move the regressor implementation from base to the regessor but that seems pretty unrelated.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -149,6 +150,32 @@ def check_estimator_sparse_data(name, Estimator):
         raise
 
 
+def check_dtype_object(name, Estimator):
```
I'd appreciate a comment here just to give a sense of what this requires of an estimator: `Numeric features must be accepted when dtype=object` for example.

amueller replied:
I'm not sure what you mean by "numeric features must be accepted" That is something that all tests require, right?
I would formulate it as "object dtype should be handled as numeric"

amueller replied:
I'm not sure what you mean by "numeric features must be accepted" That is something that all tests require, right?
I would formulate it as "object dtype should be handled as numeric"

jnothman commented:
This LGTM apart from those minor things.

GaelVaroquaux commented on sklearn/utils/validation.py:
```diff
@@ -266,11 +270,19 @@ def check_array(array, accept_sparse=None, dtype=None, order=None, copy=False,
         if ensure_2d:
             array = np.atleast_2d(array)
+        if dtype == "numeric":
+            if getattr(array, "dtype", None) is np.dtype(object):
```
Same thing here: I think that checking dtype.kind is more robust, in particular to future evolutions of numpy.

amueller commented:
rebased, used ``dtype.kind == "O"``.

amueller committed:
more robust check for dtype object

amueller commented:
Should be good now.

amueller commented:
@ogrisel merge?

ogrisel commented:
Merged!

amueller commented:
Thanks :) That should close some issues.
